orttraining/orttraining/training_ops/cuda/optimizer/adam.cu

```diff
@@ -69,6 +70,66 @@ __global__ void _AdamOptimizer(
   moment_2_out[id] = m2o;
 }
 
+template <typename T1, typename T3, typename T4, typename T_GRAD, typename T_GRAD_NORM>
+__global__ void _AdamOptimizer_mode1(
+    const T1* eta,
+    const T3* weights,
+    const T_GRAD* grads,
+    const T4* moment_1,
+    const T4* moment_2,
+    const T3* loss_scale,
+    const T_GRAD_NORM* grad_norm,
+    const T4 alpha,
+    const T4 beta,
+    const T4 lambda,
+    const T4 epsilon,
+    const T4 alpha_correction,
+    const T4 beta_correction,
+    T4* moment_1_out,
+    T4* moment_2_out,
+    T3* weights_out,
+    T_GRAD* grads_out,
+    half* fp16_weights_out,
+    CUDA_LONG N) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
+  const T4 actual_scale = _ComputeGradScale<T3, T_GRAD_NORM, T4>(loss_scale, grad_norm);
+
+  // Gradient scaling/clipping.
+  const T4 g = T4(grads[id]) / actual_scale;
+  // A shared constant.
+  const T4 one = T4(1.0f);
+
+  // Compute exponentially-averaged historical gradient.
+  const T4 m1o = alpha * moment_1[id] + (one - alpha) * g;
+
+  // Compute exponentially-averaged historical squared gradient.
+  const T4 m2o = beta * moment_2[id] + (one - beta) * g * g;
+
+  const T4 denom = _Sqrt(m2o) + epsilon;
+
+  // Apply bias correction terms on learning rate
+  const T4 step_size = T4(*eta) * _Sqrt(beta_correction) / alpha_correction;
+
+  // Huggingface updates weights in the following logic:
+  // param' = param - step_size * m1o / denom
+  // param_out = param' - original_lr * lambda * param'
+  // then param_out = param - step_size * m1o / denom - original_lr * lambda * (param - step_size * m1o / denom)
+  // so delta = -step_size * m1o / denom - original_lr * lambda * (param - step_size * m1o / denom)
+  const T4 delta = -step_size * m1o / denom - T4(*eta) * lambda * (T4(weights[id]) - step_size * m1o / denom);
```

Comment:
>; [](start = 109, length = 1)

  // Compute the new gradient.
  if (grads_out) {
    grads_out[id] = T_GRAD(delta);
  }

still need this for parasail

Reply:
thanks for catching this.. deleted by mistake.